caas/kubernetes/provider/customresourcedefinitions.go

```diff
@@ -104,24 +104,59 @@ func (k *kubernetesClient) getCustomResourceDefinition(name string) (*apiextensi
 	return crd, errors.Trace(err)
 }
 
-func (k *kubernetesClient) deleteCustomResourceDefinitions(appName string) error {
+func (k *kubernetesClient) listCustomResourceDefinitions(labels map[string]string) ([]apiextensionsv1beta1.CustomResourceDefinition, error) {
+	listOps := v1.ListOptions{
+		LabelSelector: labelsToSelector(labels),
+	}
+	list, err := k.extendedCient().ApiextensionsV1beta1().CustomResourceDefinitions().List(listOps)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(list.Items) == 0 {
+		return nil, errors.NotFoundf("custom resource definitions with labels %v", labels)
+	}
+	return list.Items, nil
+}
+
+func (k *kubernetesClient) deleteCustomResourceDefinitionsForApp(appName string) error {
+	return errors.Trace(k.deleteCustomResourceDefinitions(k.getCRDLabelsGlobal(appName)))
+}
+
+func (k *kubernetesClient) deleteCustomResourceDefinitions(labels map[string]string) error {
 	err := k.extendedCient().ApiextensionsV1beta1().CustomResourceDefinitions().DeleteCollection(&v1.DeleteOptions{
 		PropagationPolicy: &defaultPropagationPolicy,
 	}, v1.ListOptions{
-		LabelSelector: labelsToSelector(k.getCRDLabels(appName)),
+		LabelSelector: labelsToSelector(labels),
 	})
 	if k8serrors.IsNotFound(err) {
 		return nil
 	}
 	return errors.Trace(err)
 }
 
-func (k *kubernetesClient) deleteCustomResources(appName string) error {
-	crds, err := k.extendedCient().ApiextensionsV1beta1().CustomResourceDefinitions().List(v1.ListOptions{})
+func (k *kubernetesClient) deleteCustomResourcesForApp(appName string) error {
+	labelsGetter := func(crd apiextensionsv1beta1.CustomResourceDefinition) map[string]string {
+		labels := k.getCRDLabelsGlobal(appName)
+		if isCRDNameSpaced(crd.Spec.Scope) {
+			labels = k.getCRDLabelsNamespaced(appName)
```

Comment:
labels in this case will just contain the app name
won't that mean that we also select and delete all the global custom resources as well?